test: setup conformance test suite

raynerljm opened the pull request:
## Problem

As part of ensuring security and quality for sgID, we would like to ensure that our SDKs are conformant to the OIDC specifications. OIDC provides a [conformance testing tool](https://www.certification.openid.net/login.html) to check whether a relying party/server adheres to the standards laid out.

So far, we have been testing our TypeScript and Python SDK manually before releases as a sanity check (Tests are outlined and documented [here](https://www.notion.so/Conformance-tests-5cf71c390aa64756905a900f7db9cef1?pvs=21). However, in an effort to improve efficiency and maintainability (especially if we decide to develop more SDKs), we would like automate these tests.

Related to #44 

## Solution

The `demo-rp` directory contains the demo app that will be interacting with the conformance tests.

- This app is based off the Next.js CSR example app under `examples/nextjs-csr`.
- Additionally, this demo app overrides the endpoints to point to the conformance tests instead of the sgID transaction server (the overriding can be found under `test/conformance/demo-rp/src/lib/sgidClient.ts`).

The `suite` directory contains the scripts that interact with the conformance test server to start the test plan and test modules, and interact with the demo app through API calls.

- `test.py` is the main test runner that defines the config of the conformance test plan and runs each test module
- `conformance.py` (modified based off this [tutorial](https://gitlab.com/openid/conformance-suite-automated-testing-tutorial)) is a wrapper over the [API calls](https://www.certification.openid.net/api-document.html) to the conformance test server and also includes the logic to interact with the locally hosted demo app
- `modules.py` specifies the test modules (in the `oidcc-client-test-plan` test plan) to run and their expected final statuses and results

### Limitations of the test suite

- Some of the test modules (i.e. `oidcc-client-test-signing-key-rotation`) require the authorization flow to be ran twice which is currently not supported with the way the test suite is built - Therefore the test modules are commented out
    - This could be resolved by customizing the script for certain test modules but I've not embarked on this unless we decide that it's worth the effort to invest more into this test suite
- Some of the test modules test for features that we do not support (i.e. Webfinger, open id config from discovery, Aggregated/Distributed claims) and are thus commented out to reduce the flakiness of the test suite (as the behaviour is less predictable for these modules)

## Screenshots

<img width="400" alt="image" src="https://github.com/opengovsg/sgid-client/assets/63008910/a12b7c21-a57f-4f66-8233-06e555d910cc">



## Tests

You can run the conformance suite by following the instructions under `test/conformance/README.md`

The tests will be run manually when needed and will not be integrated into our CI/CD pipeline because of the following reasons (which are also outlined in the README)
>   1.  The tests take up to a few minutes to run which could be disruptive to our releases
>   2.  The tests depend on the OIDC conformance test servers and we do not want the availability of those to influence our ability to make releases
>  3.  The tests are flaky as they depend on making requests to the OIDC conformance test servers which could also be disruptive to our releases

**New environment variables**:

- `CONFORMANCE_TOKEN` : permanent token from OpenID certification to interact with the conformance suite server

raynerljm committed:
feat: setup conformance test suite

kwajiehao commented on test/conformance/README.md:
```diff
@@ -0,0 +1,77 @@
+
+To ensure that our SDKs conform to the OIDC standards, we will use a demo app that utilizes our SDK to act as an RP against the conformance tests.
+
+## How it works
```
nit: despite being labelled "How it works", this section doesn't actually explain how it works. can we include one sentence on how the conformance test relies on OpenID Foundation's server, which acts as an IdP, to run various tests on the RP software, and how the tests verify that the RP software is compliant? (for e.g. by checking the response from the RP)

raynerljm replied:
good catch, thanks! done in [7c69dc3](https://github.com/opengovsg/sgid-client/pull/60/commits/7c69dc329992e00b49f6050dd26659d9af3d87fb)

mantariksh commented on test/unit/mocks/constants.ts:
since we are moving all unit test-related files under `test/unit`, should we move the top-level `jest.config.js` and `jest.setup.js` into `test/unit` as well? we'll need to change the `test` script to point `jest` to the new location of the config file

raynerljm replied:
yeah that makes sense - resolved in [f008af6](https://github.com/opengovsg/sgid-client/pull/60/commits/f008af62e29f51214c1e081790851ea60dca3796)

mantariksh commented on test/conformance/demo-rp/package.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "nextjs-csr",
```
we can update this

raynerljm replied:
okay updated in [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh commented on test/conformance/demo-rp/public/skyline.png:
we probably don't need these images right? adds unnecessary bloat to the repo

raynerljm replied:
yeah i agree - removed them in [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh commented on test/conformance/README.md:
````diff
@@ -0,0 +1,77 @@
+     ```
+     to setup a Python virtual env
+   - Run `cat .env.example > .env` and repalce the `CONFORMANCE_TOKEN` with your [conformance testing permanent token](https://www.certification.openid.net/tokens.html) to setup your environment variables
+   - Run `pip install -r requirements.txt` to install the requirements
````
have we considered running these tests as part of our CI/CD pipeline? if we've decided against it, might be good to document why, probably in the PR description or in this README

raynerljm replied:
thanks for the comment - documented in the README here [aebdb25](https://github.com/opengovsg/sgid-client/pull/60/commits/aebdb25a185fd56e589696002ad511374a8e03b7)

mantariksh reviewed (commented):
small initial comments before I read through the rest

mantariksh commented on test/conformance/README.md:
````diff
@@ -0,0 +1,77 @@
+     python3 -m venv .venv
+     ```
+     to setup a Python virtual env
+   - Run `cat .env.example > .env` and repalce the `CONFORMANCE_TOKEN` with your [conformance testing permanent token](https://www.certification.openid.net/tokens.html) to setup your environment variables
````
```suggestion
   - Run `cat .env.example > .env` and replace the `CONFORMANCE_TOKEN` with your [conformance testing permanent token](https://www.certification.openid.net/tokens.html) to setup your environment variables
```

raynerljm replied:
thanks for the catch! resolved in [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh commented on test/conformance/README.md:
```diff
@@ -0,0 +1,77 @@
+
+1. Build the sgID SDK
+   - Run `npm run build` in the root directory (i.e. `/sgid-client`)
+1. Run the demo app locally
```
```suggestion
2. Run the demo app locally
```

raynerljm replied:
when formatted the markdown actually auto-increments ordered lists for you

i.e. this becomes
```
1. hey
1. hey
```
this
1. hey
1. hey

but I do agree that it makes it confusing when reading it in markdown - resolved in [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh commented on test/conformance/README.md:
```diff
@@ -0,0 +1,77 @@
+1. Run the demo app locally
+   - Run `cd test/conformance/demo-rp` to move to the `demo-rp` directory
+   - Run `npm i` to install the dependencies
+1. Run the test runner script
```
```suggestion
3. Run the test runner script
```

raynerljm replied:
reason for this was discussed above - resolved in [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh commented on test/conformance/README.md:
```diff
@@ -0,0 +1,77 @@
+## Running the tests
+
+1. In the `test/conformance/demo-rp` directory, run `npm run dev`
+1. In a new terminal in the `test/conformance/suite` directory
```
```suggestion
2. In a new terminal in the `test/conformance/suite` directory
```

raynerljm replied:
reason for this was discussed above - resolved in [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh commented on test/conformance/README.md:
```diff
@@ -0,0 +1,77 @@
+1. In a new terminal in the `test/conformance/suite` directory
+   - If you have not activate your Python venv, run `source .venv/bin/activate`
+   - Run `python test.py`
+1. Observe the results in the terminal
```
```suggestion
3. Observe the results in the terminal
```

raynerljm replied:
reason for this was discussed above - resolved in [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh reviewed (commented):
could you say more about why you chose the NextJS app as the demo RP? I'm concerned about us maintaining an entire full-stack NextJS app for a test, especially in a year's time when NextJS (probably) will no longer be in fashion

mantariksh commented on test/conformance/demo-rp/package.json:
```diff
@@ -0,0 +1,34 @@
+    "lint": "next lint"
+  },
+  "dependencies": {
+    "@opengovsg/sgid-client": "github:opengovsg/sgid-client#test/oidc-conformance",
```
we can update this to the published version right?

raynerljm replied:
thanks for the catch - this made me realize that our example apps were not updated as well. resolved here [07384b4](https://github.com/opengovsg/sgid-client/pull/60/commits/07384b48542897e25bfe12c4cc003551d3369b0b)

mantariksh commented on test/conformance/README.md:
```diff
@@ -0,0 +1,77 @@
+1. In the `test/conformance/demo-rp` directory, run `npm run dev`
+1. In a new terminal in the `test/conformance/suite` directory
+   - If you have not activate your Python venv, run `source .venv/bin/activate`
+   - Run `python test.py`
```
I ran this and got the following:

```
Failed modules: 3 / 25
['oidcc-client-test-discovery-openid-config', 'oidcc-client-test-signing-key-rotation-just-before-signing', 'oidcc-client-test-signing-key-rotation']
```

is that expected?

raynerljm replied:
sorry that i wasn't terribly clear about the expected behaviour. in this change [a1dab4f](https://github.com/opengovsg/sgid-client/pull/60/commits/a1dab4f2130deefdf5a0213bc442965f1dbaf651), I've added code-level comments to the `modules.py` file with the name, description, and expected behaviour (+ comments) for each test module

I've also commented out the test modules that are
1. not supported by sgID (i.e. Webfinger, open id config from discovery, Aggregated/Distributed claims)
2. not currently supported with the way the test suite is ran (i.e.`oidcc-client-test-signing-key-rotation`)

I've updated the PR description with the above-mentioned limitations of the test suite 

so right now all the tests should be passing

mantariksh commented on test/conformance/demo-rp/src/pages/api/userinfo.ts:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  // Request user info with access token
+  const { data } = await sgidClient.userinfo({
```
does the decryption always fail because the test payload isn't encrypted? might be good to reflect that in a comment

raynerljm replied:
the OIDC server returns an empty object for the userinfo payload 
<img width="448" alt="image" src="https://github.com/opengovsg/sgid-client/assets/63008910/d33d886b-b95d-4da3-b483-d2b3093b872a">

as a result, the `decryptPayload` function will not be called during execution of the test suite.

here is a test that proves that it is not actually called

modified SDK code:
<img width="578" alt="image" src="https://github.com/opengovsg/sgid-client/assets/63008910/2ff5c427-0e8a-4f20-af24-de81fa77bcc9">

server logs:
<img width="424" alt="image" src="https://github.com/opengovsg/sgid-client/assets/63008910/37108f3c-71fc-43b9-ae33-c5bcb3afe552">

thanks for the comment though! I'll make a note of this in the README + make a code level comment about this behaviour

raynerljm replied:
added the comments in the README and code level here [aebdb25](https://github.com/opengovsg/sgid-client/pull/60/commits/aebdb25a185fd56e589696002ad511374a8e03b7)

mantariksh reviewed (commented):
overall, as I said before, tbh I'm still not convinced that the payoff-to-maintainability ratio is worth it for this, for a few reasons:

1. the conformance tests are not doing anything special that we can't do ourselves in unit tests, e.g. testing for invalid sub, invalid signature. the fact that they are published by the OpenID Foundation does not make them magical.
2. the conformance tests are opinionated (e.g. enforcing discovery, which is optional as per spec), and as a result we have to determine which tests are okay to fail based on the spec anyway. we might as well just write our own unit tests by referring to the spec itself.
3. the conformance tests require a lot of code to maintain, including a whole full stack web app, test driver code and monkey-patching of our own SDK, as seen in this PR. if the tests break in a year's time, it will be hard for us to figure out what's wrong.
4. the tests are extremely costly in terms of time, as they take 6.5 minutes to run. our CI currently takes 30-40 seconds.

will leave the final decision to jiehao!

raynerljm committed:
refactor: move jest config files into test/unit

raynerljm committed:
fix: point example and demo apps to sdk in npm

additionally removed images from conformance testing demo app to remove bloat

additionally fixed typos in the conformance testing README

raynerljm commented:
> could you say more about why you chose the NextJS app as the demo RP? I'm concerned about us maintaining an entire full-stack NextJS app for a test, especially in a year's time when NextJS (probably) will no longer be in fashion

i chose the Next.js app (more specifically the CSR one) as the demo RP for the following reasons
1. i wanted to utilize one of the example apps in the `examples` directory. this is so that if we had to update the example apps (due to a change in the SDK), the additional effort required to update the demo RP for the conformance test would be minimal
2. among the example apps, the Next.js apps can mimic the full E2E RP flow (FE to BE) from a single server (compared to the express example which requires spinning up an additional server to serve the frontend files as well.

this is why the implementation of the demo RP is essentially just the Next.js CSR example app but with a modified `lib/sgidClient.ts` file to interact with the conformance test server instead.

to facilitate the discussion, do you have any suggestions for a demo RP that could potentially be easier to maintain?

raynerljm commented:
@mantariksh 

> could you say more about why you chose the NextJS app as the demo RP? I'm concerned about us maintaining an entire full-stack NextJS app for a test, especially in a year's time when NextJS (probably) will no longer be in fashion

i chose the Next.js app (more specifically the CSR one) as the demo RP for the following reasons
1. i wanted to utilize one of the example apps in the `examples` directory. this is so that if we had to update the example apps (due to a change in the SDK), the additional effort required to update the demo RP for the conformance test would be minimal
2. among the example apps, the Next.js apps can mimic the full E2E RP flow (FE to BE) from a single server (compared to the express example which requires spinning up an additional server to serve the frontend files as well.

this is why the implementation of the demo RP is essentially just the Next.js CSR example app but with a modified `lib/sgidClient.ts` file to interact with the conformance test server instead.

to facilitate the discussion, do you have any suggestions for a demo RP that could potentially be easier to maintain?

raynerljm committed:
docs: update README with reasons not to include in CI/CD pipeline

additionally add comments to README and `userinfo` endpoint about how the decryption of the payload will not be attempted

mantariksh commented:
> to facilitate the discussion, do you have any suggestions for a demo RP that could potentially be easier to maintain?

another couple of options (not mutually exclusive):
1. Go framework-free for the frontend. Use vanilla JS and just have one login button which does a `fetch` to get the auth URL. So we don't have to maintain a whole React app for the tests.
2. Use Express for the backend and have an endpoint to serve static files.

agree that NextJS does more stuff for you, but imo it'll be harder to maintain in future as NextJS and React keep upgrading. Express may upgrade too but if we just use it for a backend + static files, we'd have very little surface area which might be affected by breaking changes.

raynerljm committed:
fix: update test modules with documentation

kwajiehao commented on test/conformance/demo-rp/src/lib/sgidClient.ts:
nit: ideally, the conformance tests are run for every SDK, which means that we will need to have a wrapper client for each language's SDK. Given that, would it make sense to locate the conformance suite test code in a separate repo, since every SDK repo will need to use it?

raynerljm replied:
yup that makes a lot of sense, i refactored the `suite` directory and placed it in its own repository [here](https://github.com/opengovsg/sgid-client-conformance-suite). I've also updated the README with instructions on how to clone the conformance suite repository. [0246a91](https://github.com/opengovsg/sgid-client/pull/60/commits/0246a9174d262279a8c4d2373f107f3f04405e71)

kwajiehao commented on test/conformance/demo-rp/src/lib/sgidClient.ts:
```diff
@@ -0,0 +1,75 @@
+  'sgid-conformance-test-secret-that-should-be-at-least-256-bits-long'
+
+// Ensure that the last path parameter matches the alias in `test/conformance/suite/test.py`
+const hostname =
```
Can we come up with a different test name so that the hostname doesn't include your name?

raynerljm replied:
yup okay, done in [7c69dc3](https://github.com/opengovsg/sgid-client/pull/60/commits/7c69dc329992e00b49f6050dd26659d9af3d87fb)

kwajiehao commented on test/conformance/demo-rp/src/lib/sgidClient.ts:
```diff
@@ -0,0 +1,75 @@
+      jwks_uri: `${hostname}/jwks`,
+    })
+
+    const any_this = this as any
```
nit: can we include a code comment on why this is necessary?

raynerljm replied:
fair point! done in [7c69dc3](https://github.com/opengovsg/sgid-client/pull/60/commits/7c69dc329992e00b49f6050dd26659d9af3d87fb)

kwajiehao commented on test/conformance/demo-rp/src/pages/api/auth-url.ts:
```diff
@@ -0,0 +1,49 @@
+export default function handler(req: NextApiRequest, res: NextApiResponse) {
+  try {
+    store.clear()
+    // sgidClientService.reset()
```
nit: remove this line if not used

raynerljm replied:
thanks for the catch - resolved in [7c69dc3](https://github.com/opengovsg/sgid-client/pull/60/commits/7c69dc329992e00b49f6050dd26659d9af3d87fb)

kwajiehao commented on test/conformance/demo-rp/src/pages/api/auth-url.ts:
```diff
@@ -0,0 +1,49 @@
+      res,
+      httpOnly: true,
+      // domain: 'http://localhost:3000',
+      // sameSite: 'none',
```
What's the rationale for deviating from the example code (`examples/nextjs-csr/src/pages/api/auth-url.ts` where `secure: true`)?

raynerljm replied:
so i found that using `secure: true` results in the test suite failing as the cookie does not get set

my suspicion is that because we are not using a real browser but are just using API calls to simulate browser interaction with the RP, it recognizes localhost as insecure and thus doesn't set the cookie

I've added a code-level comment in [7c69dc3](https://github.com/opengovsg/sgid-client/pull/60/commits/7c69dc329992e00b49f6050dd26659d9af3d87fb) to explain why this change was made

kwajiehao commented on test/conformance/demo-rp/src/pages/api/callback.ts:
nit: can we standardize naming of the file with the examples (`nextjs-csr` example is called `redirect.ts`)

raynerljm replied:
mm makes sense - [7c69dc3](https://github.com/opengovsg/sgid-client/pull/60/commits/7c69dc329992e00b49f6050dd26659d9af3d87fb)

kwajiehao commented on test/conformance/suite/modules.py:
Awesome stuff! thanks for explaining the expected results

kwajiehao commented on test/conformance/suite/modules.py:
```diff
@@ -0,0 +1,567 @@
+      "status":[
+         "FINISHED"
+      ],
+      "result":[
```
nit: we can rename `result` to `expected-result` to more clearly communicate the intention of having this attribute in the dictionary

raynerljm replied:
good idea! renamed in [7c69dc3](https://github.com/opengovsg/sgid-client/pull/60/commits/7c69dc329992e00b49f6050dd26659d9af3d87fb)

kwajiehao reviewed (approved):
lgtm, left some nit comments

kwajiehao commented:
> overall, as I said before, tbh I'm still not convinced that the payoff-to-maintainability ratio is worth it for this, for a few reasons:
> 
> 1. the conformance tests are not doing anything special that we can't do ourselves in unit tests, e.g. testing for invalid sub, invalid signature. the fact that they are published by the OpenID Foundation does not make them magical.
> 2. the conformance tests are opinionated (e.g. enforcing discovery, which is optional as per spec), and as a result we have to determine which tests are okay to fail based on the spec anyway. we might as well just write our own unit tests by referring to the spec itself.
> 3. the conformance tests require a lot of code to maintain, including a whole full stack web app, test driver code and monkey-patching of our own SDK, as seen in this PR. if the tests break in a year's time, it will be hard for us to figure out what's wrong.
> 4. the tests are extremely costly in terms of time, as they take 6.5 minutes to run. our CI currently takes 30-40 seconds.

Personally I think that it's worth it to maintain these conformance tests. The main reason for doing so is to offload the burden of maintaining the test cases to OpenID Foundation. Quick response to the points you've raised:
1. We can certainly write our own unit tests but these are harder to maintain because we need to do them for each language we have a SDK in. By deferring to the OpenID Foundation test server, we just need to provide a monkey patch for the sgID client.
2. As you've pointed out, we don't actually need to update the test app - we just need to make sure that the sgID client can run the test cases. As long as we freeze our dependencies for the test app, there should be no breakage due to dependency updates and any failed tests will be cases that we should rightly investigate.
3. Since we don't release often, a few minutes for tests is not a heavy cost to pay

Also keen to hear your thoughts @raynerljm !

raynerljm committed:
doc: update README and add code-level comments

Additionally renamed `callback.ts` to `redirect.ts` to follow the Next.js CSR example.

Also renamed `result` to `expected-result`.

Updated README to explain how the test suite works

raynerljm committed:
refactor: refactor the conformance test suite into its own repository

raynerljm commented:
i was previously more on the fence on whether we should merge these tests as I wasn't sure whether it was wise to bog down the `sgid-client` repository with all the conformance-suite related code.

however, after refactoring the suite code into its own [repository](https://github.com/opengovsg/sgid-client-conformance-suite) and leaving the monkey patch code in the relevant SDK, I believe its fine to continue with the merge + with the maintenance of the conformance suite. in the event that we do find that the cost is not worth the benefit, we can just stop using the suite and leave the repository alone instead of having to delete this portion in every single SDK.

kwajiehao commented on test/conformance/README.md:
```diff
@@ -0,0 +1,40 @@
+   - Run `pip install -r requirements.txt` to install the requirements
+   - Run `cat .env.example > .env` and replace the `CONFORMANCE_TOKEN` with your [conformance testing permanent token](https://www.certification.openid.net/tokens.html) to setup your environment variables
+5. Additional notes
+   - Please ensure that the value of `test_plan_config.alias` in `test/conformance/suite/test.py` is the same as the last path parameter in `hostname` in `test/conformance/demo-rp/src/lib/sgidClient.ts` (e.g. `sgid-sdk-conformance-test`)
```
This refers to a file which no longer exists - `test/conformance/suite/test.py`

raynerljm replied:
thanks for the catch! 🎉  - updated in [71be529](https://github.com/opengovsg/sgid-client/pull/60/commits/71be5294e6de11787475ce71e47a5b5fe4fd27cc)

kwajiehao reviewed (approved):
lgtm

instructions on the [new repo](https://github.com/opengovsg/sgid-client-conformance-suite) are a bit confusing because they reference files and directories in another repo (for e.g. `test/conformance/demo-rp`) but we can deal with that later

left a nit comment on docs

raynerljm committed:
docs: update conformance suite README directory name
